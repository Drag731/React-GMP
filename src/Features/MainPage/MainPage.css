.main-page {
  max-width: 1000px;
  margin: 0 auto; }
  .main-page__header {
    padding: 10px 30px;
    position: relative; }
    .main-page__header:before {
      content: "";
      background: url("https://pluggedin.ru/images/avengers-infinity-war-mash-up-mcu-marvel-2018-1024x576-bigTopImage.jpg");
      background-size: 100% 100%;
      top: 0;
      left: 0;
      bottom: 0;
      right: 0;
      position: absolute;
      z-index: -2; }
    .main-page__header:after {
      content: "";
      background: grey;
      background-size: 100% 100%;
      top: 0;
      left: 0;
      bottom: 0;
      right: 0;
      position: absolute;
      z-index: -1;
      opacity: 0.3; }
  .main-page__result {
    padding: 10px 30px;
    background-color: lightgrey; }

.b-header__logo-wrap {
  color: red;
  font-weight: bold; }

.b-header__search {
  color: white; }
  .b-header__search-title {
    display: block;
    margin-bottom: 20px;
    text-transform: uppercase;
    font-size: 18px; }
  .b-header__search-input {
    width: 100%;
    color: white;
    background: black;
    border: 0;
    border-bottom: 2px solid red;
    padding: 10px 20px; }

.b-header__choice {
  color: white;
  padding: 10px 0;
  text-transform: uppercase; }

.b-header__button {
  background: #ccc;
  pointer-events: auto; }
  .b-header__button:hover {
    cursor: pointer; }

.b-header__button_active {
  background: aqua;
  pointer-events: none; }

.choice-flex {
  display: flex;
  justify-content: space-between;
  align-items: center; }
  .choice-flex__item:first-child {
    display: flex;
    width: 210px;
    justify-content: space-between;
    align-items: center; }
  .choice-flex__item button:hover {
    outline: none; }

.b-result {
  display: flex;
  justify-content: space-between;
  align-items: center; }
  .b-result__sort-wrap {
    display: flex;
    width: 210px;
    justify-content: space-between;
    align-items: center; }
  .b-result__sort-item {
    color: #000;
    pointer-events: auto; }
    .b-result__sort-item:hover {
      cursor: pointer; }
  .b-result__sort-item_active {
    color: red;
    pointer-events: none; }

.movie-items {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between; }

.b-item {
  max-width: 300px;
  margin-top: 30px; }
  .b-item__img-wrap {
    margin-bottom: 10px; }
  .b-item__img-pic {
    width: 100%; }
  .b-item__title-wrap {
    display: flex;
    justify-content: space-between;
    align-items: center; }
  .b-item__title-text {
    color: black;
    font-weight: bold; }
  .b-item__title-year {
    border: 1px solid black;
    border-radius: 5px;
    padding: 3px; }
  .b-item__ganre-wrap {
    color: darkgray; }
