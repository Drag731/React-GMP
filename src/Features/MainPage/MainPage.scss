.main-page {
  max-width: 1000px;
  margin:0 auto;
  &__header {
	  padding: 10px 30px;
	  background-color: pink;
	}
	&__result {
		padding: 10px 30px;
		background-color: lightgrey;
	}
}

.b-header {
	&__logo-wrap {
		color: red;
		font-weight: bold;
	}
	&__search {
		color: white;
		&-title {
			display: block;
			margin-bottom: 20px;
			text-transform: uppercase;
			font-size: 18px;
		}
		&-input {
			width: 100%;
			color: white;
			background: black;
			border: 0;
			border-bottom: 2px solid red;
			padding: 10px 20px;
		}
	}
	&__choice {
		color: white;
		padding: 10px 0;
		text-transform: uppercase;
	}
}

.choice-flex {
		display: flex;
		justify-content: space-between;
		align-items: center;
		&__item {
			&:first-child {
				display: flex;
				width: 210px;
				justify-content: space-between;
				align-items: center;
			}
			button {
				&:hover {
					outline: none;
				}
			}
		}
}

.b-result {
	display: flex;
	justify-content: space-between;
	align-items: center;
	&__sort-wrap {
		display: flex;
		width: 210px;
   		justify-content: space-between;
		align-items: center;
	}
}

.movie-items {
	display: flex;
	flex-wrap: wrap;
	justify-content: space-between;
}

.b-item {
	max-width: 300px;
	margin-top: 30px;
	&__img {
		&-wrap {
			margin-bottom: 10px;
		}
		&-pic {
			width: 100%;
		}
	}
	&__title {
		&-wrap {
			display: flex;
			justify-content: space-between;
			align-items: center;
		}
		&-text {
			color: black;
			font-weight: bold;
		}
		&-year {
			border: 1px solid black;
			border-radius: 5px;
			padding: 3px;
		}
	}
	&__ganre {
		&-wrap {
			color: darkgray;
		}
	}
}
