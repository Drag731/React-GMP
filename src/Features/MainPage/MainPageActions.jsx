import { call, put, all, takeLatest, takeEvery } from 'redux-saga/effects';
import { getMovies }  from '../../api/movies'
export const ADD_MOVIES = 'ADD_MOVIES';
export const GO_TO_DESCRIPTION = 'GO_TO_DESCRIPTION';
export const FETCH_MOVIES = 'FETCH_MOVIES';

export const fetchMovies = (query) => ({
    type: FETCH_MOVIES,
    query
});

export const receiveMoviesState = payload => ({
    type: ADD_MOVIES,
    payload,
});

export const goToDescription = () => ({
    type: GO_TO_DESCRIPTION,
});

// Sagas
export function* receiveMoviesDB({query}) {
    const response = yield call(getMovies, query);

    yield put(receiveMoviesState(response));
}

export function* watchFetchMovies() {
    yield takeLatest('FETCH_MOVIES', receiveMoviesDB);
}

export function* moviesSaga() {
    yield all([
        watchFetchMovies(),
    ]);
}