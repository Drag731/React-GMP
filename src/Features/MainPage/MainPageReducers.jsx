import {
    ADD_MOVIES,
} from './MainPageActions';

const initialState = {
    movies: [],
    total: 0,
    isLoadingMovies: true
};

const MoviesReducer = (state = initialState, action) => {
    switch (action.type) {

        case ADD_MOVIES: {
            return {
                ...state,
                isLoadingMovies: false,
                movies: action.response.data,
                total: action.response.total
            };
        }

        default: {
            return state;
        }
    }
};

export const getIsLoadingMovies = state => state.movies.isLoadingMovies;
export const getData = state => state.movies.movies;
export const getTotal = state => state.movies.total;

export default MoviesReducer;