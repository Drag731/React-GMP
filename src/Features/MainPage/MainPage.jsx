// @flow
import * as React from 'react';
import { List, Record } from 'immutable';
import { getMovies } from '../../api/movies'

import MovieItemsBlock from '../../components/MovieItemsBlock/MovieItemsBlock';
import HeaderBlock from './components/HeaderBlock/HeaderBlock'

import './MainPage.css'

type Props = { };

const mainPageRecord = Record(({
    id: 0,
    poster_path: '',
    title: '',
    release_date: 0,
    genres: []
}: {
    id: number,
    poster_path: string,
    title: string,
    release_date: number,
    genres: Array<string>
}));

const mainPageRecordInstance = mainPageRecord();

type State = {
    movies: List<typeof mainPageRecordInstance>,
    total: number,
    search: string,
    ratingSort: boolean,
    dateSort: boolean,
    isLoadingMovies: boolean
};

class MainPage extends React.Component<Props, State> {
    constructor(props: {}) {
        super(props);

        this.state = {
            movies: List(),
            total: 0,
            search: "",
            ratingSort: true,
            dateSort: true,
            isLoadingMovies: true
        };

        this.loadMovies = this.loadMovies.bind(this);
        this.handleSearch = this.handleSearch.bind(this);
        this.handleTitleSearch = this.handleTitleSearch.bind(this);
        this.handleDirectorSearch = this.handleDirectorSearch.bind(this);
        this.handleSearchButton = this.handleSearchButton.bind(this);
    }

    componentDidMount() {
        this.loadMovies(this.state.search);
    }

    loadMovies: (q: string) => void;
    loadMovies(q: string) {
        return getMovies(q)
            .then((res: {
                data: {
                    id: number,
                    poster_path: string,
                    title: string,
                    release_date: number,
                    genres: Array<string>
                }[],
                total: number
            }) => {
                return (
                    this.setState({
                        movies: List(res.data),
                        total: res.total
                    })
                )
            })
            .then(() => {
                this.setState({
                    isLoadingMovies: false
                });
            })
            .catch(() => {
                throw new Error('I crashed!');
            })
    }

    handleSearch: (event: {target: {value: string}}) => void;
    handleSearch(event: {target: {value: string}}) {
        this.setState({
            search: event.target.value
        });
    }

    handleSearchButton: () => void;
    handleSearchButton() {
        this.loadMovies(`?search=${this.state.search}&searchBy=title`)
    }

    handleTitleSearch: () => void;
    handleTitleSearch() {
        if (!this.state.search) return;
        this.loadMovies(`?search=${this.state.search}&searchBy=title`)
    }

    handleDirectorSearch: () => void;
    handleDirectorSearch() {
        if (!this.state.search) return;
        const genres = this.state.search[0].toUpperCase() + this.state.search.slice(1);
        this.loadMovies(`?search=${genres}&searchBy=genres`)
    }

    render() {

        return (
                <React.Fragment>
                    <div className="main-page container">
                        <HeaderBlock
                            handleSearch={this.handleSearch}
                            handleTitleSearch={this.handleTitleSearch}
                            handleDirectorSearch={this.handleDirectorSearch}
                            handleSearchButton={this.handleSearchButton}
                            search={this.state.search}
                        />
                        {this.state.isLoadingMovies ?
                            "Loading..." :
                            <MovieItemsBlock movies={this.state.movies}/>
                        }
                    </div>
                </React.Fragment>
        )
    }
}

export default MainPage;