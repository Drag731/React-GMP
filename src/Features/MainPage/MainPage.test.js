import React from 'react';
import renderer from 'react-test-renderer';
import {shallow, mount, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import MainPage from './MainPage.jsx'

configure({ adapter: new Adapter() });

describe('<MainPage  />', () => {
    it('Renders MainPage component', () => {
        const tree = shallow(
            <MainPage />
        );

        expect(tree.find('.main-page')).toHaveLength(1);
    });
    it('should call loadMovies on init', () => {
        const spy = jest.spyOn(MainPage.prototype, 'loadMovies');
        const tree = shallow(
            <MainPage />
        );

        expect(spy).toHaveBeenCalled();
    });
    it('should call handleReleaseDateSort on click', () => {
        const spy = jest.spyOn(MainPage.prototype, 'handleReleaseDateSort');
        const tree = mount(
            <MainPage />
        );

        tree.find('.b-result__sort-item').first().simulate('click');

        expect(spy).toHaveBeenCalled();
    });
    it('should call handleRatingSort on click', () => {
        const spy = jest.spyOn(MainPage.prototype, 'handleRatingSort');
        const spy2 = jest.spyOn(MainPage.prototype, 'loadMovies');
        const tree = mount(
            <MainPage />
        );

        tree.find('.b-result__sort-item').at(1).simulate('click');

        expect(spy).toHaveBeenCalled();
        expect(spy2).toHaveBeenCalled();
    });
    it('should call handleTitleSearch on click', () => {
        const spy = jest.spyOn(MainPage.prototype, 'handleTitleSearch');
        const spy2 = jest.spyOn(MainPage.prototype, 'loadMovies');
        const tree = mount(
            <MainPage />
        );

        tree.find('button').at(0).simulate('click');

        expect(spy).toHaveBeenCalled();
        expect(spy2).toHaveBeenCalled();
    });
    it('should call handleDirectorSearch on click', () => {
        const spy = jest.spyOn(MainPage.prototype, 'handleDirectorSearch');
        const spy2 = jest.spyOn(MainPage.prototype, 'loadMovies');
        const tree = mount(
            <MainPage />
        );

        tree.find('button').at(1).simulate('click');

        expect(spy).toHaveBeenCalled();
        expect(spy2).toHaveBeenCalled();
    });
    it('should call handleSearch on click', () => {
        const spy = jest.spyOn(MainPage.prototype, 'handleSearch');
        const spy2 = jest.spyOn(MainPage.prototype, 'loadMovies');
        const tree = mount(
            <MainPage />
        );

        tree.find('.b-header__search-input').simulate('change');

        expect(spy).toHaveBeenCalled();
        expect(spy2).toHaveBeenCalled();
    });
});
