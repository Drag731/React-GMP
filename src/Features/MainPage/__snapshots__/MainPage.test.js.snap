// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainPageTest  /> Renders DescriptionPageTest component snapshot2 1`] = `
<ErrorBoundary>
  <React.Fragment>
    <div
      className="main-page container"
    >
      <Connect(FilterBlock)
        goToDescription={[MockFunction]}
        isLoadingMovies={false}
        location={
          Object {
            "search": "a",
          }
        }
        movies={Array []}
        receiveMoviesDB={
          [MockFunction] {
            "calls": Array [
              Array [
                "a",
              ],
              Array [
                "a",
              ],
            ],
          }
        }
        total={0}
      />
      <NoFilmsFound />
    </div>
  </React.Fragment>
</ErrorBoundary>
`;

exports[`<MainPageTest  /> Renders DescriptionPageTest component snapshot3 1`] = `
<ErrorBoundary>
  <React.Fragment>
    <div
      className="main-page container"
    >
      <Connect(FilterBlock)
        goToDescription={[MockFunction]}
        isLoadingMovies={false}
        location={
          Object {
            "search": "a",
          }
        }
        movies={Array []}
        receiveMoviesDB={
          [MockFunction] {
            "calls": Array [
              Array [
                "a",
              ],
              Array [
                "a",
              ],
              Array [
                "a",
              ],
            ],
          }
        }
        total={3000}
      />
      <MovieItemsBlock
        handleGoToDescriptionPage={[Function]}
        movies={Array []}
      />
    </div>
  </React.Fragment>
</ErrorBoundary>
`;

exports[`<MainPageTest  /> Renders MainPageTest component 1`] = `
<ErrorBoundary>
  <React.Fragment>
    <div
      className="main-page container"
    >
      <Connect(FilterBlock)
        goToDescription={[MockFunction]}
        isLoadingMovies={true}
        location={
          Object {
            "search": "a",
          }
        }
        movies={Array []}
        receiveMoviesDB={
          [MockFunction] {
            "calls": Array [
              Array [
                "a",
              ],
            ],
          }
        }
        total={0}
      />
      Loading...
    </div>
  </React.Fragment>
</ErrorBoundary>
`;
